### PR TITLE
DQN of dynamic graph version

### PR DESCRIPTION
**Pull request by rical730**
1. DQN in cartPole
2. add unittest for deepcopy(model) or deepcopy(actor_critic)

### CODE REVIEW

**Comment by rical730** [comment]
TODO：check paddle version for every fluid example

**Comment by rical730** [comment]
> TODO：check paddle version for every fluid example

done in https://github.com/PaddlePaddle/PARL/pull/531

File: benchmark/paddle/DQN/replay_memory.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
+#   Copyright (c) 2020 PaddlePaddle Authors. All Rights Reserved.
```
Comment (TomorrowIsAnOtherDay) [inline benchmark/paddle/DQN/replay_memory.py]: Can we reuse the rpm in `parl.utils`?
https://github.com/PaddlePaddle/PARL/blob/develop/parl/utils/replay_memory.py

File: benchmark/paddle/DQN/README.md [hunk 2]
```diff
@@ -0,0 +1,27 @@
+## Reproduce DQN with PARL
+Based on PARL, we provide a simple demonstration of DQN.
```
Comment (TomorrowIsAnOtherDay) [inline benchmark/paddle/DQN/README.md]: we provide a simple demonstration of DQN.
----------->
we provide a simple example of the DQN algorithm.

File: benchmark/paddle/DQN/README.md [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+<img src="../QuickStart/performance.gif" alt="result" height="175"/>
+<img src="cartpole.jpg" alt="result" height="175"/>
+</p>
+
+## How to use
+### Dependencies:
++ [paddlepaddle>=2.0.0](https://github.com/PaddlePaddle/Paddle)
++ [parl>=1.4.2](https://github.com/PaddlePaddle/PARL)
++ gym
++ tqdm
```
Comment (TomorrowIsAnOtherDay) [inline benchmark/paddle/DQN/README.md]: Do we need tqdm?

File: benchmark/paddle/DQN/cartpole_agent.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+        """
+        if self.global_step % self.update_target_steps == 0:
+            self.alg.sync_target()
+        self.global_step += 1
+
+        act = np.expand_dims(act, axis=-1)
+        reward = np.expand_dims(reward, axis=-1)
+        terminal = np.expand_dims(terminal, axis=-1)
+
+        obs = paddle.to_tensor(obs.astype(np.float32))
```
Comment (TomorrowIsAnOtherDay) [inline benchmark/paddle/DQN/cartpole_agent.py]: Why do we need `xxx.as_type(xxx)` here?
Is is possible to remove them?

File: benchmark/paddle/DQN/cartpole_agent.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+        act = paddle.to_tensor(act.astype(np.int32))
+        reward = paddle.to_tensor(reward.astype(np.float32))
+        next_obs = paddle.to_tensor(next_obs.astype(np.float32))
+        terminal = paddle.to_tensor(terminal.astype(np.float32))
+        loss = self.alg.learn(obs, act, reward, next_obs, terminal)
+        return loss.numpy()[0]
```
Comment (TomorrowIsAnOtherDay) [inline benchmark/paddle/DQN/cartpole_agent.py:101]: Please cast the paddle tensor to np.array in the `algorithm` scope rather than the `agent` scope.

File: benchmark/paddle/DQN/cartpole_agent.py [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+        self.act_dim = act_dim
+
+        self.global_step = 0
+        self.update_target_steps = 200
+
+        self.e_greed = e_greed
+        self.e_greed_decrement = e_greed_decrement
+
+    def sample(self, obs):
+        """Sample an action when given an observation
```
Comment (TomorrowIsAnOtherDay) [inline benchmark/paddle/DQN/cartpole_agent.py]: Sample an action `for exploration` when given an observation

File: benchmark/paddle/DQN/cartpole_agent.py [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+            act = np.random.randint(self.act_dim)
+        else:
+            if np.random.random() < 0.01:
+                act = np.random.randint(self.act_dim)
+            else:
+                act = self.predict(obs)
+        self.e_greed = max(0.01, self.e_greed - self.e_greed_decrement)
+        return act
+
+    def predict(self, obs):
```
Comment (TomorrowIsAnOtherDay) [inline benchmark/paddle/DQN/cartpole_agent.py:59]: Predict an action `for evaluation` when given an observation

File: benchmark/paddle/QuickStart/README.md [hunk 8]
```diff
@@ -4,7 +4,7 @@ Train an agent with PARL to solve the CartPole problem, a classical benchmark in
 ## How to use
 ### Dependencies:
 
-+ [parl](https://github.com/PaddlePaddle/PARL)
++ [parl>=1.4](https://github.com/PaddlePaddle/PARL)
```
Comment (TomorrowIsAnOtherDay) [inline benchmark/paddle/QuickStart/README.md:7]: 1.4.2

File: parl/algorithms/paddle/dqn.py [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+__all__ = ['DQN']
+
+
+class DQN(parl.Algorithm):
+    def __init__(self, model, gamma=None, lr=None):
+        """ DQN algorithm
+
+        Args:
+            model (parl.Model): model defining forward network of Q function.
+            gamma (float): discounted factor for reward computation.
```
Comment (TomorrowIsAnOtherDay) [inline parl/algorithms/paddle/dqn.py]: discounted factor for `accumulative` reward computation

File: parl/algorithms/paddle/dqn.py [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+
+__all__ = ['DQN']
+
+
+class DQN(parl.Algorithm):
+    def __init__(self, model, gamma=None, lr=None):
+        """ DQN algorithm
+
+        Args:
+            model (parl.Model): model defining forward network of Q function.
```
Comment (TomorrowIsAnOtherDay) [inline parl/algorithms/paddle/dqn.py]: forward neural network representing the Q function.

File: parl/algorithms/paddle/dqn.py [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+        assert isinstance(lr, float)
+        self.gamma = gamma
+        self.lr = lr
+
+        self.mse_loss = paddle.nn.MSELoss(reduction='mean')
+        self.optimizer = paddle.optimizer.Adam(
+            learning_rate=lr, parameters=self.model.parameters())
+
+    def predict(self, obs):
+        """ use value model self.model to predict the action value
```
Comment (TomorrowIsAnOtherDay) [inline parl/algorithms/paddle/dqn.py]: use self.model (Q function) to predict the action values

File: parl/algorithms/paddle/dqn.py [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+        """
+        return self.model.value(obs)
+
+    def learn(self, obs, action, reward, next_obs, terminal):
+        """ update value model self.model with DQN algorithm
```
Comment (TomorrowIsAnOtherDay) [inline parl/algorithms/paddle/dqn.py]: update the Q function (self.model) with DQN algorithm

File: benchmark/paddle/DQN/README.md (hunk 2 above)
Comment (rical730) [reply benchmark/paddle/DQN/README.md]: done

File: benchmark/paddle/DQN/README.md (hunk 3 above)
Comment (rical730) [reply benchmark/paddle/DQN/README.md]: removed

File: benchmark/paddle/DQN/cartpole_agent.py (hunk 7 above)
Comment (rical730) [reply benchmark/paddle/DQN/cartpole_agent.py:59]: done

File: benchmark/paddle/DQN/cartpole_agent.py (hunk 6 above)
Comment (rical730) [reply benchmark/paddle/DQN/cartpole_agent.py]: done

File: benchmark/paddle/DQN/cartpole_agent.py (hunk 4 above)
Comment (rical730) [reply benchmark/paddle/DQN/cartpole_agent.py]: replaced by `, dtype=xxxx`

File: benchmark/paddle/DQN/cartpole_agent.py (hunk 5 above)
Comment (rical730) [reply benchmark/paddle/DQN/cartpole_agent.py:101]: replaced by `, dtype=xxxx`

File: benchmark/paddle/DQN/replay_memory.py (hunk 1 above)
Comment (rical730) [reply benchmark/paddle/DQN/replay_memory.py]: deleted

File: benchmark/paddle/QuickStart/README.md (hunk 8 above)
Comment (rical730) [reply benchmark/paddle/QuickStart/README.md:7]: policy gradient can be import in parl 1.4

File: parl/algorithms/paddle/dqn.py (hunk 10 above)
Comment (rical730) [reply parl/algorithms/paddle/dqn.py]: done

File: parl/algorithms/paddle/dqn.py (hunk 9 above)
Comment (rical730) [reply parl/algorithms/paddle/dqn.py]: done

File: parl/algorithms/paddle/dqn.py (hunk 11 above)
Comment (rical730) [reply parl/algorithms/paddle/dqn.py]: done

File: parl/algorithms/paddle/dqn.py (hunk 12 above)
Comment (rical730) [reply parl/algorithms/paddle/dqn.py]: done

**Review by TomorrowIsAnOtherDay — APPROVED** [review]
LGTM.
